src/net/tcp/nbd.c

```diff
@@ -0,0 +1,1095 @@
+ */
+static void nbd_start_tx ( struct nbd_session *nbd,
+			   enum nbd_tx_state tx_state ) {
+	assert ( nbd->tx_state == NBD_TX_IDLE );
+
+	/* Set new TX state */
+	nbd->tx_state = tx_state;
+
+	/* Start transmission process */
+	nbd_tx_resume ( nbd );
+}
+
+/**
+ * Transmit an option (NBD_OPT_EXPORT_NAME or NBD_OPT_GO)
+ *
+ * @v nbd		NBD session
+ * @ret rc		Return status code
+ */
+int nbd_tx_neg_opt ( struct nbd_session *nbd ) {
+	struct io_buffer *iobuf;
+	struct {
+		uint32_t client_flags;
+		union {
+			struct nbd_proto_opt_export_name opt_export_name;
+			struct nbd_proto_opt_go opt_go;
+		};
+	} __attribute__ (( packed )) *opt;
+	size_t name_len, opt_len;
+	int rc;
+
+	name_len = strlen ( nbd->export_name );
+	if ( nbd->use_opt_go )
+		opt_len = offsetof ( typeof ( *opt ),
+				     opt_go.export_name ) +
+				     name_len +
+				     sizeof ( uint16_t );
+	else
+		opt_len = offsetof ( typeof ( *opt ),
+				     opt_export_name.export_name ) +
+				     name_len;
+
+	iobuf = xfer_alloc_iob ( &nbd->socket, opt_len );
+	if ( ! iobuf )
+		return -ENOMEM;
+
+	opt = iob_put ( iobuf, opt_len );
+	memset ( opt, 0, opt_len );
+
+	opt->client_flags =
+		htonl ( NBD_FLAG_C_FIXED_NEWSTYLE |
+			( nbd->handshake_flags & NBD_FLAG_NO_ZEROES ?
+			NBD_FLAG_C_NO_ZEROES : 0 ) );
+
+	if ( nbd->use_opt_go ) {
+		/* Construct NBD_OPT_GO option with empty NBD_INFO list */
+		opt->opt_go.request.request_magic =
+			htonll ( NBD_OPT_REQ_MAGIC );
+		opt->opt_go.request.option = htonl ( NBD_OPT_GO );
+		opt->opt_go.request.length = htonl ( opt_len -
+			offsetof ( typeof ( *opt ), opt_go.name_length ) );
+		opt->opt_go.name_length = htonl ( name_len );
+		strcpy ( opt->opt_go.export_name, nbd->export_name );
+	} else {
+		/* Construct NBD_OPT_EXPORT_NAME option */
+		opt->opt_export_name.request.request_magic =
+			htonll ( NBD_OPT_REQ_MAGIC );
+		opt->opt_export_name.request.option =
+			htonl ( NBD_OPT_EXPORT_NAME );
+		opt->opt_export_name.request.length = htonl ( name_len );
+		strcpy ( opt->opt_export_name.export_name, nbd->export_name );
+	}
+
+	DBGCIO ( nbd, "transmit:\n" );
+	DBGCIO_HD ( nbd, opt, opt_len );
+
+	/* Deliver packet */
+	if ( ( rc = xfer_deliver_iob ( &nbd->socket,
+				       iob_disown ( iobuf ) ) ) != 0 ) {
+		DBGC ( nbd, "NBD %p cannot transmit: %s\n",
+		       nbd, strerror ( rc ) );
+		return rc;
+	}
+
+	nbd->rx_state = nbd->use_opt_go ?
+		NBD_RX_NEG_OPT_INFO : NBD_RX_NEG_EXP_NAME;
+	nbd->tx_state = NBD_TX_IDLE;
+	return 0;
+}
+
+/**
+ * Process the block_read_capacity command
+ *
+ * @v nbd		NBD session
+ */
+void nbd_tx_cmd_brc ( struct nbd_session *nbd ) {
+	struct nbd_command *nbdcmd = nbd->command;
+	struct block_device_capacity capacity;
+
+	/* Sanity check */
+	assert ( nbdcmd != NULL && nbdcmd->type == NBD_BLOCK_READ_CAPACITY );
+
+	capacity.blocks = nbd->export_size / NBD_BLOCK_SIZE;
+	capacity.blksize = NBD_BLOCK_SIZE;
+	/* Use reasonable amount of data */
+	capacity.max_count = 16 * 1024 / NBD_BLOCK_SIZE;
```

Comment:
Set to `-1U` (as done in `scsi.c`) if there is no fixed limit.

Reply:
The SCSI does not enforce a limit, but iSCSI itself divides the PDU into 8192 resp. 512 bytes. NBD does not have an internal division into PDUs. I think it is good to keep the limit on the maximum size of network requests.

Reply:
The NBD_MAX_BLOCK_COUNT used